### PR TITLE
fix: swap lat and lng

### LINKED ISSUE #2587: 🐛 [Bug] Geolocation. lat and lng are swapped
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-12-19T21:43:08Z" title="Thursday, December 19th 2019, 10:43:08 pm +01:00">Dec 19, 2019</time>_
_Closed <time datetime="2020-02-18T16:55:58Z" title="Tuesday, February 18th 2020, 5:55:58 pm +01:00">Feb 18, 2020</time>_
---

When selecting a location in the user profile page, the latitude and the longitude are exchanged when saved in the database.

### Steps to reproduce the behavior
1. Go to your profile
2. Select Configuration
3. Type in some City or Region
4. Save
5. Check the lat and lng properties in the location node
grafik

### Expected behavior
The value of lat should be in lng and reverse.

### PR DESCRIPTION
**Pull request by Tirokk**
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-12-20T00:58:57Z" title="Friday, December 20th 2019, 1:58:57 am +01:00">Dec 20, 2019</time>_
_Merged <time datetime="2020-02-18T16:55:58Z" title="Tuesday, February 18th 2020, 5:55:58 pm +01:00">Feb 18, 2020</time>_
---

### Issues
- fixes #2587

### Todo
- [x] Find a solution how to repair the wrong data in production. Best would be to have script that calls the data by id from the geomap api for each location and merges the received data correctly in the lng and lat fields.

### CODE REVIEW

**Comment by Tirokk** [comment]
> [cypress[bot]](https://github.com/apps/cypress) **Authored by [cypress[bot]](https://github.com/apps/cypress)**
_<time datetime="2019-12-20T01:23:13Z" title="Friday, December 20th 2019, 2:23:13 am +01:00">Dec 20, 2019</time>_

---



<p align="center">
  <br />
  
</p>

---

### Test summary
<a href="https://dashboard.cypress.io/projects/qa7fe2/runs/5257/specs"></a> <b title="Passed">61</b>  &bull; <a href="https://dashboard.cypress.io/projects/qa7fe2/runs/5257/failures"></a> <b title="Failed">0</b>  &bull;  <b title="Pending">0</b> &bull;  <b title="Skipped">0</b>

---

### Run details
| | |
| --- | --- |
| **Project** | Human-Connection |
| **Status** | Passed |
| **Commit** | c25329f6f9 |
| **Started** | Feb 18, 2020 1:49 PM |
| **Ended** | Feb 18, 2020 2:02 PM |
| **Duration** | 13:24 <a href="https://on.cypress.io/parallelization" title="Enable parallelization to reduce run duration">💡</a> |
| **OS** | Linux Ubuntu Linux - 16.04 |
| **Browser** | Electron 78  |

[View run in Cypress Dashboard ➡️](https://dashboard.cypress.io/projects/qa7fe2/runs/5257/specs)



---

This comment has been generated by **cypress-bot** as a result of this project's GitHub integration settings. You can manage this integration in [this project's settings in the Cypress Dashboard](https://dashboard.cypress.io/projects/qa7fe2/settings)

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-07T15:00:06Z" title="Tuesday, January 7th 2020, 4:00:06 pm +01:00">Jan 7, 2020</time>_

---


before deploying this, we need to do a database manipulation where we swap the existing locations, as @Mogge said in our pairing session.

This script works... 
```
MATCH (location:Location) 
WITH location.lng as longitude, location.lat as latitude, location 
SET location.lng = latitude, location.lat = longitude 
RETURN location;
```

**Comment by Tirokk** [comment]
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2020-01-07T15:01:03Z" title="Tuesday, January 7th 2020, 4:01:03 pm +01:00">Jan 7, 2020</time>_

---


```cypher
MATCH (l:Location) WHERE NOT(l.lat IS NULL)
WITH l.lng AS longitude, l.lat AS latitude, l AS Location
SET Location.lat = longitude, Location.lng = latitude
```

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-07T15:04:07Z" title="Tuesday, January 7th 2020, 4:04:07 pm +01:00">Jan 7, 2020</time>_

---


Why do we have `Location` with no `lat` property @Mogge ??

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        SET location.lat = longitude, location.lng = latitude
+    `)
+    await transaction.commit()
+    next()
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error)
+    await transaction.rollback()
+    // eslint-disable-next-line no-console
+    console.log('rolled back')
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:29]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-02-11T11:50:18Z" title="Tuesday, February 11th 2020, 12:50:18 pm +01:00">Feb 11, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,55 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description =
+  'This migration swaps the value stored in Location.lat with the value of Location.lng.'
+
+export async function up(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+
+  try {
+    // Implement your migration here.
+    await transaction.run(`
+        MATCH (l:Location) WHERE NOT(l.lat IS NULL)
+        WITH l.lng AS longitude, l.lat AS latitude, l AS Location
+        SET Location.lat = longitude, Location.lng = latitude
+    `)
+    await transaction.commit()
+    next()
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error)
+    await transaction.rollback()
+    // eslint-disable-next-line no-console
+    console.log('rolled back')
```


```suggestion
    console.log('rolled back')
    throw new Error(error)
```

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description = `
+This migration swaps the value stored in Location.lat with the value
+of Location.lng. This is necessary as the values of lat and lng were
+stored incorrectly. For example Hamburg, Germany, was stored with the
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:6]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-02-11T11:52:46Z" title="Tuesday, February 11th 2020, 12:52:46 pm +01:00">Feb 11, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,55 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description =
+  'This migration swaps the value stored in Location.lat with the value of Location.lng.'
+
+export async function up(next) {
```


```suggestion
const swap = (next) => {
```

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    await transaction.commit()
+    next()
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error)
+    await transaction.rollback()
+    // eslint-disable-next-line no-console
+    console.log('rolled back')
+    throw new Error(error)
+  } finally {
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:31]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-02-11T11:53:15Z" title="Tuesday, February 11th 2020, 12:53:15 pm +01:00">Feb 11, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,55 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description =
+  'This migration swaps the value stored in Location.lat with the value of Location.lng.'
+
+export async function up(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+
+  try {
+    // Implement your migration here.
+    await transaction.run(`
+        MATCH (l:Location) WHERE NOT(l.lat IS NULL)
+        WITH l.lng AS longitude, l.lat AS latitude, l AS Location
+        SET Location.lat = longitude, Location.lng = latitude
+    `)
+    await transaction.commit()
+    next()
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error)
+    await transaction.rollback()
+    // eslint-disable-next-line no-console
+    console.log('rolled back')
+  } finally {
+    session.close()
+  }
+}
+
+export async function down(next) {
```


Instead:
```js
export async function up(next) {
  swap(next)
}
export async function down(next) {
  swap(next)
}
```

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js (hunk 1 above)
Comment (Tirokk) [reply backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:29]: > [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2020-02-12T13:58:50Z" title="Wednesday, February 12th 2020, 2:58:50 pm +01:00">Feb 12, 2020</time>_

---


Done

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js (hunk 2 above)
Comment (Tirokk) [reply backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:6]: > [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2020-02-12T13:59:04Z" title="Wednesday, February 12th 2020, 2:59:04 pm +01:00">Feb 12, 2020</time>_

---


Done

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:31]: > [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2020-02-12T13:59:14Z" title="Wednesday, February 12th 2020, 2:59:14 pm +01:00">Feb 12, 2020</time>_

---


Done

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description = `
+This migration swaps the value stored in Location.lat with the value
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:4]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-14T11:31:12Z" title="Friday, February 14th 2020, 12:31:12 pm +01:00">Feb 14, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,37 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description =
+  'This migration swaps the value stored in Location.lat with the value of Location.lng.'
```


I would love a better description for why this was necessary, especially for our future selves/other programmers to understand why you would ever want to swap these values

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+stored incorrectly. For example Hamburg, Germany, was stored with the
+values lat=10.0 and lng=53.55, which is close to the horn of Africa,
+but it is lat=53.55 and lng=10.0
+`
+
+const swap = async function(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+  try {
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:15]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-14T11:33:08Z" title="Friday, February 14th 2020, 12:33:08 pm +01:00">Feb 14, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,37 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description =
+  'This migration swaps the value stored in Location.lat with the value of Location.lng.'
+
+const swap = async function(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+  try {
+    // Implement your migration here.
+    await transaction.run(`
+        MATCH (l:Location) WHERE NOT(l.lat IS NULL)
+        WITH l.lng AS longitude, l.lat AS latitude, l AS Location
+        SET Location.lat = longitude, Location.lng = latitude
```


I prefer to use lower case for variables so as not to be confused with nodes... at first I thought you were trying to use set to set the node properties directly

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js (hunk 1 above)
Comment (Tirokk) [reply backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:29]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-14T11:33:53Z" title="Friday, February 14th 2020, 12:33:53 pm +01:00">Feb 14, 2020</time>_

---


I like this! could you please update the template as well, so we don't need to give this review feedback every time?

File: backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    // Implement your migration here.
```
Comment (Tirokk) [inline backend/src/db/migrations/20200206190233-swap_latitude_with_longitude.js:16]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-18T16:50:00Z" title="Tuesday, February 18th 2020, 5:50:00 pm +01:00">Feb 18, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,42 @@
+import { getDriver } from '../../db/neo4j'
+
+export const description = `
+This migration swaps the value stored in Location.lat with the value
+of Location.lng. This is necessary as the values of lat and lng were
+stored incorrectly. For example Hamburg, Germany, was stored with the
+values lat=10.0 and lng=53.55, which is close to the horn of Africa,
+but it is lat=53.55 and lng=10.0
+`
+
+const swap = async function(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+  try {
+    // Implement your migration here.
```


```suggestion
    
```
I would remove this line, but I won't request changes...

**Review by Mogge — CHANGES_REQUESTED** [review]
interrupt

If the migration fails you should either:
* throw the error again
* or call `next(error)` with error not being null

That way, the migration will not be marked as successful and the other migrations won't run.

**Review by Mogge — CHANGES_REQUESTED** [review]
some thoughts I had... really cool you caught this! not sure what happened!

**Review by Mogge — APPROVED** [review]
Looks good to me now, can we merge?

**Review by Mogge — APPROVED** [review]

**Pull request closed without merging**